skll/learner.py

```diff
@@ -1732,12 +1734,55 @@ def predict(self, examples, prediction_prefix=None, append=False,
         # Need to do some transformations so the features are in the right
         # columns for the test set. Obviously a bit hacky, but storing things
         # in sparse matrices saves memory over our old list of dicts approach.
-        if isinstance(self.feat_vectorizer, FeatureHasher):
-            if (self.feat_vectorizer.n_features !=
-                    examples.vectorizer.n_features):
+
+        # We also need to think about the various combinations of the model
+        # vectorizer and the vectorizer for the set for which we want to make
+        # predictions:
+
+        # 1. Both vectorizers are DictVectorizers. If they use different sets
+        # of features, we raise a warning and transform the features of the
+        # prediction set from its space to the trained model space.
+
+        # 2. Both vectorizers are FeatureHashers. If they use different number
+        # of feature bins, we should just raise an error since there's no
+        # inverse_transform() available for a FeatureHasher - the hash function
+        # is not reversible.
+
+        # 3. The model vectorizer is a FeatureHasher but the prediction feature
+        # set vectorizer is a DictVectorizer. We should be able to handle this
+        # case, since we can just call inverse_transform() on the DictVectorizer
+        # and then transform() on the FeatureHasher?
+
+        # 4. The model vectorizer is a DictVectorizer but the prediction feature
+        # set vectorizer is a FeatureHasher. Again, we should raise an error here
+        # since there's no inverse available for the hasher.
+        model_is_dict = isinstance(self.feat_vectorizer,
+                                   (DictVectorizer, OldDictVectorizer))
+        model_is_hasher = isinstance(self.feat_vectorizer, FeatureHasher)
+        data_is_dict = isinstance(examples.vectorizer,
+                                  (DictVectorizer, OldDictVectorizer))
+        data_is_hasher = isinstance(examples.vectorizer, FeatureHasher)
+
+        both_dicts = model_is_dict and data_is_dict
+        both_hashers = model_is_hasher and data_is_hasher
+        model_hasher_and_data_dict = model_is_hasher and data_is_dict
+        model_dict_and_data_hasher = model_is_dict and data_is_hasher
+
+        # 1. both are DictVectorizers
+        if both_dicts:
+            if (set(self.feat_vectorizer.feature_names_) !=
+                    set(examples.vectorizer.feature_names_)):
                 self.logger.warning("There is mismatch between the training model "
                                     "features and the data passed to predict.")
```

Comment:
Should this warning also indicate how the mismatch will be resolved?

"The prediction features will be traansformed to the trained model space."

Reply:
Good idea. I will do that!